add intrinsic to produce a `DigestEntries` from a `Digest` which reference file content by hash

tdyas opened the pull request:
## Motivation

The Go and JVM plugins both have a need to create a `Digest` dynamically from files from another `Digest`, but without needing to actually manipulate the content bytes of the files in the `Digest`. The engine's Filesystem API, however, only provides the `FileContent` and `DigestContents` types for obtaining information about file content, and the `CreateDigest` intrinsic only accepts `FileContent` and `Directory` instances. This means that both plugins are asking the engine to materialize file content into memory that will not ever be inspected by Python code. (Indeed, this is what the plugins are currently doing as a work around despite the performance problems.)

## Solution

Introduce the `FileEntry` type which is similar to `FileContent` but refers to the file's content using a `FileDigest` instead of including the file's actual content bytes. The PR also adds a `DigestEntries` type which can be requested via a new intrinsic to obtain the collection of `Union[FileEntry, Directory]` objects for a `Digest`. (Files are represented by `FileEntry`, and empty leaf directories by `Directory` in a similar manner to `CreateDigest`.)

The `CreateDigest` intrinsic is updated to accept `FileEntry` instances along with already-accepted `FileContent` and `Directory` instances.

[ci skip-build-wheels]

tdyas commented:
cc @patricklaw

stuhood commented:
> The Go and JVM plugins both have a need to create a Digest dynamically from files from another Digest, but without needing to actually manipulate the content bytes of the files in the Digest.

Could you include pointers to the usecases for this? Is it effectively being used to rename files?

patricklaw commented:
> > The Go and JVM plugins both have a need to create a Digest dynamically from files from another Digest, but without needing to actually manipulate the content bytes of the files in the Digest.
> 
> Could you include pointers to the usecases for this? Is it effectively being used to rename files?

In the case of Coursier fetch, it's being used to create a reproducible lockfile from a `Digest` obtained by pulling specific files from the Coursier cache into the Pants artifact DB.  Relevant rule where this happens here: https://github.com/pantsbuild/pants/blob/1552119cef6fecd10ed26d41a621a001c8c82058/src/python/pants/jvm/resolve/coursier_fetch.py#L176-L198

tdyas commented:
Cross-reference: I added links to examples in the applicable issue: https://github.com/pantsbuild/pants/issues/11806#issuecomment-908632409

stuhood commented on src/rust/engine/fs/store/src/lib.rs:
```diff
@@ -1231,6 +1231,47 @@ impl Store {
+  ) -> BoxFuture<'static, Result<Vec<FileEntry>, String>> {
+    self
+      .walk(digest, move |_, path_so_far, _, directory| {
+        future::try_join_all(
+          directory
+            .files
+            .iter()
+            .map(|file_node| {
+              let path = path_so_far.join(file_node.name.clone());
+              let is_executable = file_node.is_executable;
+              let file_node_digest = try_future!(require_digest(file_node.digest.as_ref()));
+              future::ready(Ok(FileEntry {
+                path,
+                digest: file_node_digest,
+                is_executable,
+              }))
+              .boxed()
+            })
+            .collect::<Vec<_>>(),
+        )
+        .boxed()
```
It doesn't look like this needs to use `try_join_all`: the body of the loop over files is creating Futures unnecessarily. So maybe just imperatively create the `FileEntry`s and then have a single `future::read(Ok(fie_entries))` at the end of this closure.

tdyas replied:
Fixed.

stuhood reviewed (approved):
Thanks!

While I think that we need a higher level API to ease the renaming/removing/filtering of files, this facility seems like it unblocks prototyping something like that in Python.

tdyas commented:
> While I think that we need a higher level API to ease the renaming/removing/filtering of files, this facility seems like it unblocks prototyping something like that in Python.

Correct. Whatever the higher-level API ends up being, this API as an intrinsic will allow implementing that API entirely in Python.

tdyas commented:
Rebased and modified the API sp that `DigestEntries` is a collection of `Union[FileEntry, Directory]` instead of being just a collection of `FileEntry`. This means that a roundtrip of a `Digest` to `DigestEntries` and back through `CreateDigest` should end up with the same `Digest`.
